### PR TITLE
feat(uid-mux): framed mux

### PR DESCRIPTION
**Pull request by sinui0**
This PR adds support for muxers which return framed streams.

# Changes
- Add `Codec` trait to `serio`.
- Add `FramedUidMux` trait to `uid-mux` and provide a simple mux wrapper which implements it.

### CODE REVIEW

**Review by th4s — APPROVED** [review]

File: uid-mux/src/serio.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+/// A multiplexer that opens framed streams with unique ids.
+#[async_trait]
+pub trait FramedUidMux<Id> {
+    /// Stream type.
+    type Framed: IoDuplex;
+    /// Error type.
+    type Error;
+
+    /// Open a new framed stream with the given id.
```
Comment (themighty1) [inline uid-mux/src/serio.rs]:
```suggestion
    /// Opens a new framed stream with the given id.
```

File: uid-mux/src/serio.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+/// A framed multiplexer.
+#[derive(Debug)]
+pub struct FramedMux<M, C> {
+    mux: M,
+    codec: C,
+}
+
+impl<M, C> FramedMux<M, C> {
+    /// Create a new `FramedMux`.
```
Comment (themighty1) [inline uid-mux/src/serio.rs]:
```suggestion
    /// Creates a new `FramedMux`.
```

File: uid-mux/src/serio.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    pub fn codec(&self) -> &C {
+        &self.codec
+    }
+
+    /// Returns a mutable reference to the codec.
+    pub fn codec_mut(&mut self) -> &mut C {
+        &mut self.codec
+    }
+
+    /// Split the `FramedMux` into its parts.
```
Comment (themighty1) [inline uid-mux/src/serio.rs]:
```suggestion
    /// Splits the `FramedMux` into its parts.
```

File: uid-mux/src/serio.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    Id: Sync,
+    M: UidMux<Id> + Sync,
+    C: Codec<<M as UidMux<Id>>::Stream> + Sync,
+{
+    /// Stream type.
+    type Framed = <C as Codec<<M as UidMux<Id>>::Stream>>::Framed;
+    /// Error type.
+    type Error = <M as UidMux<Id>>::Error;
+
+    /// Open a new framed stream with the given id.
```
Comment (themighty1) [inline uid-mux/src/serio.rs]:
```suggestion
    /// Opens a new framed stream with the given id.
```

File: uid-mux/src/test_utils.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+use tokio::io::{duplex, DuplexStream};
+use tokio_util::compat::{Compat, TokioAsyncReadCompatExt};
+use yamux::{Config, Mode};
+
+use crate::{
+    yamux::{Yamux, YamuxCtrl},
+    FramedMux,
+};
+
+/// Create a test pair of yamux instances.
```
Comment (themighty1) [inline uid-mux/src/test_utils.rs]:
```suggestion
/// Creates a test pair of yamux instances.
```

File: uid-mux/src/test_utils.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+) -> (Yamux<Compat<DuplexStream>>, Yamux<Compat<DuplexStream>>) {
+    let (a, b) = duplex(buffer);
+
+    let a = Yamux::new(a.compat(), Config::default(), Mode::Client);
+    let b = Yamux::new(b.compat(), Config::default(), Mode::Server);
+
+    (a, b)
+}
+
+/// Create a test pair of framed yamux instances.
```
Comment (themighty1) [inline uid-mux/src/test_utils.rs]:
```suggestion
/// Creates a test pair of framed yamux instances.
```

**Review by themighty1 — APPROVED** [review]
ack

**Pull request merged**